File: LeetCode/Medium/Count Complete Tree Nodes/Cpp/count_complete_tree_nodes.cc
#include <iostream>
#include <queue>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

typedef std::queue<TreeNode*> qt;

int count_nodes(TreeNode *root) {
    if (root == nullptr) return 0;
    qt q;
    int cnt = 0;
    q.push(root);
    while(!q.empty()) {
        auto curr = q.front();
        q.pop();
        cnt++;
        if (curr->left) q.push(curr->left);
        if (curr->right) q.push(curr->right);
    }
    return cnt;
}

int main() {
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    root->right->left = new TreeNode(6);
    std::cout<<count_nodes(root)<<"\n";
}

File: LeetCode/Medium/Count Complete Tree Nodes/Go/count_complete_tree_nodes.go
package main

import "fmt"

type TreeNode struct {
	Val int
	Left,
	Right *TreeNode
}

func New(val int) *TreeNode { return &TreeNode{Val: val} }

func count_complete_tree_nodes(root *TreeNode) int {
	cnt := 0
	if root == nil {
		return cnt
	}
	q := []*TreeNode{root}
	for len(q) > 0 {
		curr := q[0]
		q = q[1:]
		cnt++
		if curr.Left != nil {
			q = append(q, curr.Left)
		}
		if curr.Right != nil {
			q = append(q, curr.Right)
		}
	}
	return cnt
}

func main() {
	root := New(1)
	root.Left = New(2)
	root.Right = New(3)
	root.Left.Left = New(4)
	root.Left.Right = New(5)
	root.Right.Left = New(6)
	fmt.Println(count_complete_tree_nodes(root))
}
